File: node/src/tools/npm.mjs
import { npmrcPath } from "../util/paths.mjs";
import { readSafe, writeAtomic } from "../util/io.mjs";
import { readKey, setKey, removeKey } from "../util/lines.mjs";
import { readExtras, applyExtras, removeExtras } from "../util/extras.mjs";
import { buildPreflight } from "../util/version.mjs";

export const name = "npm";
export const key = "min-release-age";
export const docs = "https://docs.npmjs.com/cli/v11/using-npm/config#min-release-age";
export const minBin = [11, 10, 0];
export const extras = [
  { key: "audit-level", expected: "high", line: "audit-level=high" },
  { key: "allow-git", expected: "root", line: "allow-git=root", safeValues: ["none", "root"] },
  { key: "allow-remote", expected: "root", line: "allow-remote=root", safeValues: ["none", "root"] },
  { key: "allow-file", expected: "root", line: "allow-file=root", safeValues: ["none", "root"] },
  { key: "allow-directory", expected: "root", line: "allow-directory=root", safeValues: ["none", "root"] },
  { key: "strict-allow-scripts", expected: "true", line: "strict-allow-scripts=true" },
  { key: "dangerously-allow-all-scripts", expected: "false", line: "dangerously-allow-all-scripts=false" },
];

export function path(ctx) { return npmrcPath(ctx.env, ctx.home); }

export const preflight = buildPreflight(name, minBin,
  "min-release-age is silently ignored. Upgrade npm to enforce the cooldown.");

export async function read(ctx) {
  const p = path(ctx);
  const raw = await readSafe(p);
  const value = readKey(raw, key);
  return {
    path: p, configured: value,
    days: value === null ? null : Number(value),
    extras: readExtras(raw, extras)
  };
}

export async function write(days, ctx) {
  const p = path(ctx);
  let text = setKey(await readSafe(p), key, `${key}=${days}`);
  text = applyExtras(text, extras);
  await writeAtomic(p, text);
  return { path: p };
}

export async function unset(ctx) {
  const p = path(ctx);
  const before = await readSafe(p);
  const cooldown = removeKey(before, key);
  const ex = removeExtras(cooldown.text, extras);
  const removed = cooldown.removed || ex.removed;
  if (removed) await writeAtomic(p, ex.text);
  return { path: p, removed };
}

File: python/src/pmsec/tools/npm.py
from __future__ import annotations

from pathlib import Path

from pmsec.util.context import Context
from pmsec.util.extras import apply_extras, read_extras, remove_extras
from pmsec.util.io import write_atomic
from pmsec.util.lines import read_key, remove_key, set_key
from pmsec.util.paths import npmrc_path
from pmsec.util.version import build_preflight

NAME = "npm"
KEY = "min-release-age"
DOCS = "https://docs.npmjs.com/cli/v11/using-npm/config#min-release-age"
MIN_BIN = (11, 10, 0)
EXTRAS = [
    {"key": "audit-level", "expected": "high", "line": "audit-level=high"},
    {"key": "allow-git", "expected": "root", "line": "allow-git=root"},
    {"key": "allow-remote", "expected": "root", "line": "allow-remote=root"},
    {"key": "allow-file", "expected": "root", "line": "allow-file=root"},
    {"key": "allow-directory", "expected": "root", "line": "allow-directory=root"},
]

preflight = build_preflight(
    NAME, MIN_BIN,
    "min-release-age is silently ignored. Upgrade npm to enforce the cooldown.",
)


def path(ctx: Context) -> Path:
    return npmrc_path(ctx.env, ctx.home)


def read(ctx: Context) -> dict:
    p = path(ctx)
    raw = p.read_text("utf-8") if p.exists() else ""
    value = read_key(raw, KEY)
    days = None if value is None else int(value)
    return {"path": str(p), "configured": value, "days": days, "extras": read_extras(raw, EXTRAS)}


def write(days: int, ctx: Context) -> dict:
    p = path(ctx)
    raw = p.read_text("utf-8") if p.exists() else ""
    text = apply_extras(set_key(raw, KEY, f"{KEY}={days}"), EXTRAS)
    write_atomic(p, text)
    return {"path": str(p)}


def unset(ctx: Context) -> dict:
    p = path(ctx)
    if not p.exists():
        return {"path": str(p), "removed": False}
    before = p.read_text("utf-8")
    after, removed_cooldown = remove_key(before, KEY)
    after, removed_extras = remove_extras(after, EXTRAS)
    removed = removed_cooldown or removed_extras
    if removed:
        write_atomic(p, after)
    return {"path": str(p), "removed": removed}
